### PR TITLE
[FLINK-27991][table-planner] ORC format supports reporting statistics

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

In FLIP-231, [https://cwiki.apache.org/confluence/display/FLINK/FLIP-231%3A+Introduce+SupportsStatisticReport+to+support+reporting+statistics+from+source+connectors](https://github.com/apache/flink/pull/url), we aims to support csv, parquet, orc format for reporting statistics. This pr is a part of it to support Orc format. The details are as follows:

- In `OrcFileFormatFactory`, for class `OrcBulkDecodingFormat`, we implements `FileBasedStatisticsReportableInputFormat` interface, and implements reportStatistics() method. In this method, we cover all types stats of Orc supported, including complex types.

- Adding two UT tests, `OrcFormatStatisticsReportTest` and `OrcFileSystemStatisticsReportTest`, they all extends from `StatisticsReportTestBase` and cover the logical of Orc stats report.

## Brief change log

- In `OrcFileFormatFactory`, for class `OrcBulkDecodingFormat`, we implements `FileBasedStatisticsReportableInputFormat` interface, and implements reportStatistics() method.
- Adding two UT tests, `OrcFormatStatisticsReportTest` and `OrcFileSystemStatisticsReportTest`.

## Verifying this change

Using two UT tests, `OrcFormatStatisticsReportTest` and `OrcFileSystemStatisticsReportTest` to verify change, including single orc file and multi orc files in folder.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper:  no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? no docs

### CODE REVIEW

File: flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    private String[] formatProperties() {
+        List<String> ret = new ArrayList<>();
+        ret.add("'format'='orc'");
+        ret.add("'orc.compress'='snappy'");
+        return ret.toArray(new String[0]);
+    }
+
+    @Test
```
Comment (reswqa) [inline flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java:101]: New test class should use junit5

File: flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for statistics functionality in {@link OrcFileFormatFactory}. */
+public class OrcFileSystemStatisticsReportTest extends TableTestBase {
+
+    private BatchTableTestUtil util;
+    private TableEnvironment tEnv;
+    private String path1;
+
+    @Before
```
Comment (reswqa) [inline flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java:61]: @BeforeEach

File: flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                                + " c * 3.14 as g, h, j"
+                                + " from originalT")
+                .await();
+        // write second parquet file.
+        tEnv.executeSql(
+                        "insert into orcFileSystemTable select a, b, c, d, e, f,"
+                                + " c * 3.14 as g, h, j"
+                                + " from originalT")
+                .await();
+        assertThat(Objects.requireNonNull(new File(path1).listFiles()).length).isEqualTo(2);
```
Comment (reswqa) [inline flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java:169]:
```suggestion
        assertThat(new File(path1).listFiles()).isNotNull().hasSize(2);
```

**Comment by reswqa** [comment]
It seems that juni5 is not used. Is it because TableTestBase is not used?

**Comment by swuferhong** [comment]
> 

Hi, Now TableTestBase not using juni5, and one of my test class extends TableTestBase. So I think the best way to upgrade to juni5 should modify TableTestBase first.

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java [hunk 4]
```diff
@@ -167,5 +188,179 @@ public ChangelogMode getChangelogMode() {
...
+                    new ColumnStats.Builder().setNdv(null).setAvgLen(null).setMaxLen(null);
+            if (!columnStatistics.hasNull()) {
+                builder.setNullCount(0L);
+            } else {
+                builder.setNullCount(totalRowCount - columnStatistics.getNumberOfValues());
+            }
+
+            switch (logicalType.getTypeRoot()) {
+                case BOOLEAN:
+                    builder.setMax(null).setMin(null);
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:293]: the default value of max, min is null, so `builder.setMax(null).setMin(null);` is no needed.

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java [hunk 5]
```diff
@@ -167,5 +188,179 @@ public ChangelogMode getChangelogMode() {
...
+                    if (columnStatistics instanceof DoubleColumnStatistics) {
+                        builder.setMax(((DoubleColumnStatistics) columnStatistics).getMaximum())
+                                .setMin(((DoubleColumnStatistics) columnStatistics).getMinimum());
+                        break;
+                    } else {
+                        return null;
+                    }
+                case CHAR:
+                case VARCHAR:
+                case VARBINARY:
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:317]: no max, min for VARBINARY type

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java [hunk 6]
```diff
@@ -167,5 +188,179 @@ public ChangelogMode getChangelogMode() {
...
+                    if (columnStatistics instanceof StringColumnStatistics) {
+                        builder.setMax(((StringColumnStatistics) columnStatistics).getMaximum())
+                                .setMin(((StringColumnStatistics) columnStatistics).getMinimum());
+                        break;
+                    } else {
+                        return null;
+                    }
+                case DATE:
+                    if (columnStatistics instanceof DateColumnStatistics) {
+                        builder.setMax(((DateColumnStatistics) columnStatistics).getMaximum())
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:327]: the `Date` type of `((DateColumnStatistics) columnStatistics).getMaximum()` is not `java.sql.Date`

File: flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        path1 = tempFolder().newFolder().toURI().getPath();
+
+        BatchTableEnvUtil.registerCollection(
+                tEnv,
+                "originalT",
+                TestData.buildInData(),
+                TestData.buildInType(),
+                "a,b,c,d,e,f,g,h,i,j");
+
+        String ddl1 =
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/test/java/org/apache/flink/orc/OrcFileSystemStatisticsReportTest.java:77]: please create the test base class for statistics report and prepare the test data for all data types in the base class

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java (hunk 6 above)
Comment (godfreyhe) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:327]: please add some test to cover the Date type change

**Review by godfreyhe — COMMENTED** [review]
Thanks for the comment, I left some comments

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java (hunk 6 above)
Comment (swuferhong) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:327]: > the `Date` type of `((DateColumnStatistics) columnStatistics).getMaximum()` is not `java.sql.Date`

done!

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java [hunk 8]
```diff
@@ -167,5 +190,187 @@ public ChangelogMode getChangelogMode() {
...
+                                (Date) ((DateColumnStatistics) columnStatistics).getMaximum();
+                        Date minimum =
+                                (Date) ((DateColumnStatistics) columnStatistics).getMinimum();
+                        builder.setMax(maximum).setMin(minimum);
+                        break;
+                    } else {
+                        return null;
+                    }
+                case TIMESTAMP_WITHOUT_TIME_ZONE:
+                case TIME_WITHOUT_TIME_ZONE:
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:344]: `TIME_WITHOUT_TIME_ZONE` should be considered alone

**Review by godfreyhe — COMMENTED** [review]
Thanks for the update

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java (hunk 8 above)
Comment (swuferhong) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:344]: > `TIME_WITHOUT_TIME_ZONE` should be considered alone

Ok. Btw, orc don't support TIMESTAMP_WITHOUT_LOCAL_TIME_ZONE.

**Comment by swuferhong** [comment]
@flinkbot run azure

File: flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java [hunk 9]
```diff
@@ -167,5 +190,181 @@ public ChangelogMode getChangelogMode() {
...
+            } else {
+                builder.setNullCount(totalRowCount - columnStatistics.getNumberOfValues());
+            }
+
+            switch (logicalType.getTypeRoot()) {
+                case BOOLEAN:
+                case ROW:
+                case ARRAY:
+                case MAP:
+                    break;
```
Comment (godfreyhe) [inline flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:298]: these unsupported types can be moved to `default` branch

[the remark this answers is not in the document]
Comment (swuferhong) [reply flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileFormatFactory.java:298]: > those unsupported types can be moved to the `default` branch, they all have `nullCount`

`boolean` is not an unsupported type, I remove others except `boolean.`

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]
LGTM

**Pull request closed without merging**